task-worklow-refactor/README.md

```diff
@@ -0,0 +1,94 @@
+# How to split one Celery task into multiple tasks
+
+Let's talk about splitting the `task.process_message` into one processing task
+and multiple tasks handling the specific events. The main benefit from this 
+will be the possibility to use more queues (for faster/slower tasks)
+ and to differentiate tasks easily (Sentry).
+
+
+## Workflow
+ - get webhook payload or message from message bus 
+ - add aditional info which can make determining the event type easier
+ (GH webhooks - X-GitHub-Event)
+ - send it to Celery as `task.process_message` - this task will take care about:
+    - parsing of the event into object of class Event (this class provides `project`, `package_config` properties,
+     which enables us to make the checks)
+    - private repository check
+    - getting the handlers which handle the event
+    - whitelist check
+    - possibly creating records in DB (explained below)
+    - (running `pre-check` here to filter out some events?)
+    - sending specific task to Celery
+  - handle specific tasks, which directly run the handlers
+    
+#### How to pass the information needed by handlers
+ Handlers are currently using the information from event (with `project` and `package_config` properties, specific for each event),
+  service config and job config object:
+ ```python
+class JobHandler(Handler):
+    def __init__(
+        self, config: ServiceConfig, job_config: Optional[JobConfig], event: Event
+    )
+```
+ Possible solutions, which can be somehow combined:
+ 1. serialize the info about objects and pass it into `send_task`
+     - this would need serializing and then again deserializing
+ 2. save the info about objects in DB and pass IDs of models into `send_task`
+     - what models does make sense to have? possibilities:
+         - project
+         - package config 
+         - service config
+         - job config 
+         - event
+     
+     - each subclass of class Event stores different set of info -> does it make sense to create model for each? 
+     - these could be reused since in `task_results` table we store the event dict
+ 3. pass just the arguments which are required by the specific handler
```

Comment:
This one will probably be the easiest one.

Reply:
This would go with nr 1, right?

I feel that going with the DB solution would lead to a more fragile solution, as the DB would become a single point of failure: no tasks—not even the ones not depending on calls to DB—could be created or processed.

Also, the serialisation problem seems to be a much easier one to solve than coming up with the right DB models.

Reply:
from my point of view #3 is best option - there will be very clean boundaries between objects.

Reply:
Yes, this will be much cleaner, but we need to forward all the information that is provided by `Event` instances. (And make some nice abstraction on this to not have a high number of arguments.)

And parts of the handlers' code is, unfortunately, event-dependent. These I can remember now:

- how to get a package config (easily forwardable)
- which project we want to work with (forwardable with some work)
- reporting (e.g. comments are nor possible for branch-push/release)
  - some abstraction should be done here (e.g. instead of event type, have `is_commenting`)

Reply:
We can maybe use also the trigger models from the database.

Reply:
I know exactly all this dependencies and by this i am suggesting to break them. It was main source of issues in CentOS/Pagure implementation.

e.g.
[CoprBuildJobHelper](https://github.com/packit-service/packit-service/blob/9e455f253f77ecda028473c49aa67b04c9927bff/packit_service/worker/build/copr_build.py#L52) 
* why it has Event dependency? In real world you can use copr api without Event, why dont do it that way? It will follow [OOP desgin principles](https://deviq.com/solid/)

* Also it contains "unrelated" methods (e.g. [self.report_status_to_all_for_chroot](https://github.com/packit-service/packit-service/blob/9e455f253f77ecda028473c49aa67b04c9927bff/packit_service/worker/build/copr_build.py#L159-L164)) which requires unrelated properties (e.g. [project](https://github.com/packit-service/packit-service/blob/9e455f253f77ecda028473c49aa67b04c9927bff/packit_service/worker/build/build_helper.py#L236-L242) ) , it is making from CoprBuildJobHelper some kind of [semi-God object](https://en.wikipedia.org/wiki/God_object)

Reply:
@sakalosj I answered most of it already here and during the arch-day.

Some notes:
- We already agreed that we want to rename and clear `BuildHelper` classes.
- `BuildHelpers` do not run build itself, it actually calls the methods on `packit.PackitAPI`. (So its purpose is to trigger build and report the status in the meantime. So, It's not so God-like.)
- Don't forget that this task was meant to be the first step -- we can't rewrite the whole service here...;) It should be doable and reviewable.

---
Off-topic: When you regularly cites the OOP principles, it feels like you consider us/me very stupid. I know them.

Reply:
@lachmanfrantisek 
sorry, didn't want to hurt anyone. It wasn't meant to attack, just to discuss my opinions, which can be wrong and this is one of platform to discuss.

Per my understanding this is research, and i think it should mentioned here. It doesn't mean it  have to be implemented all at once.

I used term "some kind of semi-God object", to express difference, that that there is something more that should be, but it is not  pure "God object".

I understand it like, [reporting](https://github.com/packit-service/packit-service/blob/9e455f253f77ecda028473c49aa67b04c9927bff/packit_service/worker/build/copr_build.py#L159-L164)) is acting upon one of git projects, therefore status reporter should be part of (or directly communicating with) class representing this git project. If we move it to BuildHelper we require also to pass project related informations. Currently it is done by adding project and config info to Event. Therefore I called it semi-God.

I think there is few places to improve, and because OOP principles is well known terminology, to cite them is easiest way to cover ideas i want to share.